src/RustyIceberg.jl

```diff
@@ -64,297 +30,600 @@ else
+            # Select columns if specified
+            if !isempty(iter.columns)
+                select_columns!(scan, iter.columns)
             end
-        end
-
-        state = IcebergTableIteratorState(table, scan, true)
-    end
 
-    # Get next batch
-    result, batch_ptr = iceberg_scan_next_batch(state.scan)
-
-    if result == ICEBERG_END_OF_STREAM
-        # End of stream - cleanup and return nothing
-        iceberg_scan_free(state.scan)
-        iceberg_table_free(state.table)
-        return nothing
-    elseif result != ICEBERG_OK
-        # Error - cleanup and throw
-        iceberg_scan_free(state.scan)
-        iceberg_table_free(state.table)
-        error("Failed to get next batch: $(iceberg_error_message())")
-    end
+            if !isnothing(iter.data_file_concurrency_limit)
+                with_data_file_concurrency_limit!(scan, iter.data_file_concurrency_limit)
+            end
 
-    if batch_ptr == C_NULL
-        error("Received NULL batch")
-    end
+            if !isnothing(iter.manifest_entry_concurrency_limit)
+                with_manifest_entry_concurrency_limit!(scan, iter.manifest_entry_concurrency_limit)
+            end
 
-    # Convert ArrowBatch pointer to ArrowBatch struct
-    batch = unsafe_load(batch_ptr)
+            if !isnothing(iter.batch_size)
+                with_batch_size!(scan, iter.batch_size)
+            end
 
-    # Create IOBuffer from the serialized Arrow data
-    io = IOBuffer(unsafe_wrap(Array, batch.data, batch.length))
+            stream = scan!(scan)
+            state = TableIteratorState(table, scan, stream, true)
+        else
+            # Clean up the batch from the previous iteration.
+            if state.batch_ptr != C_NULL
+                free_batch(state.batch_ptr)
+                state.batch_ptr = C_NULL
+            end
+        end
 
-    # Read Arrow data
-    arrow_table = Arrow.Table(io)
+        # Wait for next batch asynchronously
+        state.batch_ptr = next_batch(state.stream)
 
-    # Free the batch
-    iceberg_arrow_batch_free(batch_ptr)
+        if state.batch_ptr == C_NULL
+            # End of stream - mark for cleanup and return nothing
+            should_cleanup_resources = true
+            return nothing
+        end
 
-    return arrow_table, state
+        # Convert ArrowBatch pointer to ArrowBatch struct
+        batch = unsafe_load(state.batch_ptr)
+
+        # Read Arrow data
+        arrow_table = Arrow.Table(unsafe_wrap(Array, batch.data, batch.length))
+
+        return arrow_table, state
+
+    catch e
+        # On exception, mark resources for cleanup before rethrowing
+        should_cleanup_resources = true
+        rethrow(e)
+    finally
+        # Clean up scan and table resources only if needed (end of stream or exception)
+        if should_cleanup_resources && state !== nothing && state.is_open
+            state.batch_ptr != C_NULL && free_batch(state.batch_ptr)
+            free_stream(state.stream)
```

Comment:
what happens if we created table, but exception was thrown before we created the state? Maybe we should create state right away in before opening things, and then assign `scan`, `table`, etc. to it. and free should check if they are `nothing` and only free if they are not.